Add new codemods package (#13884)
microsoft/fluentui: packages/codemods/README.md

# ts-morphin-migration

> ts-morphin-migration is a package created and managed with the `create-just` utility.

Reviewer: Can you update this with the current package name and description? (remove the reference to create-just unless you did actually use that, and add a real description)

> ## Next Steps

Reviewer 2: It would be great to start the README.md to explain what the package is for and how it should be used, rather than how to test it.

Run

```sh
yarn
yarn start-test
```

> To start testing the codemods

Reviewer: In general I think the description of what the commands do should go above the commands not below

## To add a codemod

Add your codemods to the `./src/mods` folder with `.mod.ts|tsx` as the file type.

## Test with the actual package:

Run

```sh
yarn build
```

> To build the upgrade.js file

Reviewer: What is the upgrade.js file?

Run

```sh
npm pack
```

> from the codemods root to create a tar file for testing. Move the created tar file to the package you want to test and run

Reviewer 3: ```suggestion
from the codemods package's root to create a tar file for testing. Move the created tar file to the package you want to test and run
```

```sh
npx <tarFileName>
```

## Executing a codemod (Prototype)

> Make sure that fluent is installed in your package.

Reviewer 3: ```suggestion
Make sure that Fluent UI is installed in your package.
```

Reviewer 2: Even then, I'm confused. What does this mean? In your "application"? What "Fluent UI" package? Why is this package separate from other packages?

E.g. if `react-button` wants to ship a mod with the package, can the mod live with the package? What pattern are we suggesting to use for partners (like react-date-time) that want to write codemods? If they are separate from the package they apply to, how are they locked to the package they apply to (how does versioning work?)

Author: Right now, for v1, this will run all codemods present in this package. There is no reason we can't change where the execution file looks to get codemods but that can be a feature down the line.

> Make sure that fluent is installed in your package.

Reviewer: You should list the specific package(s) here. I assume it works with fui/react but also add OUFR if it works with that.
```suggestion
Make sure that `@fluentui/react` is installed as a dependency of your package.
```

Author: Right now the codemods don't have a concept of package or root. That's probably a good thing to add as a filter or next steps.

Run

```

> npx @fluentui\codemods

Reviewer: This should probably use a forward slash?

> npx @fluentui\codemods

Reviewer 4: just want to confirm these will be supported:
1. User can pick a specific codemod to run, e.g: repath to compat path
2. User can target a specific component in case of repathing

Author: I don't believe that the goal is to support this for V1. We can talk more about it, however.

Reviewer 4: I left this as an Ask in hackmd. we can discuss further with the team. regardless, this is something outside of the scope of this PR

```

## Todos

- Write a `flag` utility that will enable devs to note when a part of a file needs to be changed, but cannot be done via codemod.
- Implement a command that will execute all listed codemods on a single file.
  - Will need to think of a way to specify the order. Maybe something like tasks in Just.

>   - This could be helpful when you want to run a set of codemods based on a single condition, like the presence of a specific import

Reviewer 3: Should these be tracked here? It feels like these should probably be tracked in an issue instead of in the README.

Author: I'm going to keep them here for now, since this is a prototype/alpha I think it's fine to have more stuff here. It will be removed before it's published.

# @fluentui/codemods

This is a utility package to assist with the upgrading of libraries and apps that rely on fluentui.

## How it works

> If you have a typescript application or library that relies on fluentui then you can run `npx @fluentui/codemods` which will immediately begin an upgrade of your codebase so that you don't have to! It works by finding all the tsconfig files and then using those to find the relevant files to upgrade before running the updates on each of them!

Reviewer 3: ```suggestion
If you have a typescript application or library that relies on a non-current version of Fluent UI then you can run `npx @fluentui/codemods` to immediately begin an upgrade of your codebase, saving you the trouble of doing so manually! This works by finding all the tsconfig files and then using those to find the relevant files to upgrade before running the updates on each of them!
```

## Executing a codemod (Prototype)

> If your application relies on any package in fluentui simply run

Reviewer 3: ```suggestion
If your application relies on any Fluent UI package simply run
```

```
npx @fluentui/codemods
```

> and the upgrade will begin if there are any relevant codemods!

Reviewer 3: ```suggestion
and the upgrade will begin if there are any relevant codemods to apply to your codebase!
```

## Test with the actual package:

Run

```sh
yarn build
```

To build

Run

```sh
npm pack
```

from the codemods package root to create a tar file for testing. Move the created tar file to the package you want to test and run

```sh
npx <tarFileName>
```

## Todos

- Write a `flag` utility that will enable devs to note when a part of a file needs to be changed, but cannot be done via codemod.
- Implement a command that will execute all listed codemods on a single file.
  - Will need to think of a way to specify the order. Maybe something like tasks in Just.
  - This could be helpful when you want to run a set of codemods based on a single condition, like the presence of a specific import

## Notes

ts-morph:

- ts-morph does most of the heavy lifting. Don't be afraid to use it directly rather than trying to abstract into a utility.
- One of the most useful types of the syntax tree to get is the `SyntaxKind.block` it is the equivalent of the `{ stuff }` that is located in a function declaration and is where a lot of code lives.
- You can only access JSX props on syntax kinds of `SyntaxKind.JSXOpeningElement` and `SyntaxKind.JSXSelfClosingElement`

> - `getChildIndex` returns the child index respective to the immediate parent. It resets each level. So consider the following:

Reviewer 3: ```suggestion
- `getChildIndex` returns the child index respective to the immediate parent. It resets at each level. So consider the following:
```

# @fluentui/codemods

This is a utility package to assist with the upgrading of libraries and apps that rely on fluentui.

## How it works

> If you have a typescript application or library that relies on fluentui then you can run `npx @fluentui/codemods` which will immediately begin an upgrade of your codebase so that you don't have to! It works by finding all the tsconfig files and then using those to find the relevant files to upgrade before running the updates on each of them!

Reviewer: I still think it would be helpful to specify the particular package names it's capable of upgrading (in particular that OUFR is supported too)

Author: Currently nothing is explicitly supported, we're still figuring out how all of that will work.
